x/bank/types/params.go

```diff
@@ -3,31 +3,141 @@ package types
 
-// ParamKeyTable type declaration for parameters
+// ParamKeyTable for bank module.
 func ParamKeyTable() paramtypes.KeyTable {
-	return paramtypes.NewKeyTable(
-		paramtypes.NewParamSetPair(ParamStoreKeySendEnabled, false, validateSendEnabled),
-	)
+	return paramtypes.NewKeyTable().RegisterParamSet(&Params{})
 }
 
-func validateSendEnabled(i interface{}) error {
-	_, ok := i.(bool)
+// NewParams creates a new parameter configuration for hte bank module
+func NewParams(sendEnabledParams SendEnabledParams) Params {
+	return Params{
+		SendEnabled: sendEnabledParams,
+	}
+}
+
+// DefaultParams is the default parameter configuration for the bank module
+func DefaultParams() Params {
+	return Params{
+		SendEnabled: SendEnabledParams{DefaultSendEnabled()},
+	}
+}
+
+// Validate all bank module parameters
+func (p Params) Validate() error {
+	if err := validateSendEnabledParams(p.SendEnabled); err != nil {
+		return err
+	}
+	return nil
+}
+
+// String implements the Stringer interface.
+func (p Params) String() string {
+	out, _ := yaml.Marshal(p)
+	return string(out)
+}
+
+// IsSendEnabled returns true if the given denom is enabled for sending
+func (p Params) IsSendEnabled(denom string) bool {
+	var defaultSendEnabled = true
+
+	for _, p := range p.SendEnabled {
+		if p.Denom == denom {
+			return p.Enabled
+		}
+		// capture default case
+		if len(p.Denom) == 0 {
+			defaultSendEnabled = p.Enabled
+		}
+	}
+
+	return defaultSendEnabled
+}
+
+// SetSendEnabledParam returns an updated set of Parameters with the given denom
+// send enabled flag set.
+func (p Params) SetSendEnabledParam(denom string, sendEnabled bool) Params {
+	var sendParams SendEnabledParams
+	for _, p := range p.SendEnabled {
+		if p.Denom != denom {
+			sendParams = append(sendParams, NewSendEnabled(p.Denom, p.Enabled))
+		}
+	}
+	sendParams = append(sendParams, NewSendEnabled(denom, sendEnabled))
+	return NewParams(sendParams)
+}
+
+// ParamSetPairs implements params.ParamSet
+func (p *Params) ParamSetPairs() paramtypes.ParamSetPairs {
+	return paramtypes.ParamSetPairs{
+		paramtypes.NewParamSetPair(KeySendEnabled, &p.SendEnabled, validateSendEnabledParams),
+	}
+}
+
+// SendEnabledParams is a collection of parameters indicating if a coin denom is enabled for sending
+type SendEnabledParams []*SendEnabled
+
+func validateSendEnabledParams(i interface{}) error {
+	params, ok := i.([]*SendEnabled)
 	if !ok {
 		return fmt.Errorf("invalid parameter type: %T", i)
 	}
+	// ensure each denom is only registered one time.
+	registered := make(map[string]bool)
+	for _, p := range params {
+		if err := validateSendEnabled(*p); err != nil {
+			return err
+		}
+		if _, exists := registered[p.Denom]; exists {
+			return fmt.Errorf("duplicate send enabled parameter found: '%s'", p.Denom)
+		}
+		registered[p.Denom] = true
+	}
+	return nil
+}
+
+// DefaultSendEnabled returns the default send enabled parameter
+func DefaultSendEnabled() *SendEnabled {
```

Comment:
I don't understand the empty denom on this function. Does it mean that every denomination has transfers enabled?

Reply:
The use of the empty denom string provides a straight forward way to configure the previous behavior of a global send_enabled flag combined with coin denomination specific overrides.  The specific logic used can be found here https://github.com/cosmos/cosmos-sdk/blob/856881953cc7877e24cdd1ce43d663f260b083dc/x/bank/types/params.go#L53

I am open to additional comments or an alternate approach if that makes sense.  This capability is something that would be useful in my specific situation that has many different denoms in use and only a small number that are disabled.

Possible alternatives:

- Additional clarification through source code comments and documentation
- Refactor this capability into an explicit parameter outside of the list of coin denomination specific parameters
- Remove this feature entirely

Reply:
I agree that it's a useful configuration. I'd personally go with option 2) and have boolean (`EnableAllSends`) that overwrites any individual enabled denomination. The empty string can be misinterpreted.

Reply:
I'm open to other options too @clevinson @aaronc

Reply:
Using the name `EnabledAllSends` may indicate behavior outside of intended scope.  The use of `All` in this context may lead to potential ambiguity with `SendEnabled` configuration that includes coins configured with the opposite behavior.  I suggest the use of the term `Default` instead here as this value will be used when no more specific configuration applies.